pilot/pkg/model/config.go

```diff
@@ -748,7 +782,33 @@ func (store *istioConfigStore) ExternalServices() []Config {
 	return configs
 }
 
+// TODO: move the logic to v2, read all VirtualServices once at startup and per
+// change event and pass them to the config generator. Model calls to List are
+// extremely expensive - and for larger number of services it doesn't make sense
+// to just convert again and again, for each listener times endpoints.
 func (store *istioConfigStore) VirtualServices(gateways []string) []Config {
+	if len(gateways) == 1 && gateways[0] == IstioIngressGatewayName {
+		configs, err := store.List(VirtualServiceIngress.Type, NamespaceAll)
+		if err != nil {
+			return nil
+		}
+		configs3, err := store.List(VirtualService.Type, NamespaceAll)
+		if err != nil {
+			return nil
+		}
+		for _, config := range configs3 {
+			rule := config.Spec.(*networking.VirtualService)
+			if len(rule.Gateways) > 0 {
+				for _, ruleGateway := range rule.Gateways {
+					if ruleGateway == IstioIngressGatewayName {
+						configs = append(configs, config)
+						break
```

Comment:
ITs not that easy. You have to deal with conflicts. And merge the rules. Thats the reason I completely avoided this piece. Doing this is essentially bringing back all the ingress bugs.